src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Services/SchemaService.cs

```diff
@@ -1,66 +1,76 @@
-using System;
+using System.Collections.Concurrent;
 using System.Collections.Generic;
 using System.Linq;
+using System.Threading;
 using System.Threading.Tasks;
 using GraphQL;
 using GraphQL.Types;
-using Microsoft.Extensions.Caching.Memory;
 using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Primitives;
+using OrchardCore.Environment.Shell.Scope;
 
 namespace OrchardCore.Apis.GraphQL.Services
 {
     public class SchemaService : ISchemaFactory
     {
-        private readonly IMemoryCache _memoryCache;
         private readonly IEnumerable<ISchemaBuilder> _schemaBuilders;
-        private readonly IServiceProvider _serviceProvider;
+        private readonly SemaphoreSlim _schemaGenerationSemaphore = new SemaphoreSlim(1, 1);
+        private readonly ConcurrentDictionary<ISchemaBuilder, IChangeToken> _changeTokens = new ConcurrentDictionary<ISchemaBuilder, IChangeToken>();
 
-        public SchemaService(
-            IMemoryCache memoryCache,
-            IEnumerable<ISchemaBuilder> schemaBuilders,
-            IServiceProvider serviceProvider)
+        private ISchema _schema;
+
+        public SchemaService(IEnumerable<ISchemaBuilder> schemaBuilders)
         {
-            _memoryCache = memoryCache;
             _schemaBuilders = schemaBuilders;
-            _serviceProvider = serviceProvider;
         }
 
-        public Task<ISchema> GetSchema()
+        public async Task<ISchema> GetSchemaAsync()
         {
-            return _memoryCache.GetOrCreateAsync("GraphQLSchema", async f =>
+            if (_schema is object && !_changeTokens.Values.Any(x => x.HasChanged))
             {
-                f.SetSlidingExpiration(TimeSpan.FromHours(1));
+                return _schema;
+            }
+
+            await _schemaGenerationSemaphore.WaitAsync();
+
+            try
+            {
+                if (_schema is object && !_changeTokens.Values.Any(x => x.HasChanged))
+                {
+                    return _schema;
+                }
+
+                var serviceProvider = ShellScope.Services;
 
-                ISchema schema = new Schema
+                var schema = new Schema
                 {
                     Query = new ObjectGraphType { Name = "Query" },
                     Mutation = new ObjectGraphType { Name = "Mutation" },
                     Subscription = new ObjectGraphType { Name = "Subscription" },
                     FieldNameConverter = new OrchardFieldNameConverter(),
-
-                    DependencyResolver = _serviceProvider.GetService<IDependencyResolver>()
+                    DependencyResolver = serviceProvider.GetService<IDependencyResolver>()
                 };
 
                 foreach (var builder in _schemaBuilders)
                 {
                     var token = await builder.BuildAsync(schema);
 
-                    if (token != null)
+                    if (token is object)
```

Comment:
`token != null` and everywhere else it's used

Reply:
@sebastienros do you want another pr with that change .. someone merged this one it seems ;p

Reply:
whats wrong with `is object` SonarQube allow it :P